tsl/src/compression/create.c

```diff
@@ -741,6 +743,201 @@ validate_hypertable_for_compression(Hypertable *ht)
 	}
 }
 
+/*
+ * Get the default segment by value for a hypertable
+ */
+static ArrayType *
+compression_setting_segmentby_get_default(const Hypertable *ht)
+{
+	StringInfo command;
+	StringInfo result;
+	int res;
+	ArrayType *column_res = NULL;
+	Datum datum;
+	text *message;
+	bool isnull;
+	MemoryContext upper = CurrentMemoryContext;
+	MemoryContext old;
+	int32 confidence = -1;
+
+	if (ts_guc_default_segmentby_fn == NULL || strlen(ts_guc_default_segmentby_fn) == 0)
+	{
+		elog(DEBUG1,
+			 "segment_by default: hypertable=\"%s\" columns=\"\" function: \"\" confidence=-1",
+			 get_rel_name(ht->main_table_relid));
+		return NULL;
+	}
+
+	command = makeStringInfo();
+	appendStringInfo(command,
+					 "SELECT "
+					 " (SELECT array_agg(x)  "
+					 " FROM jsonb_array_elements_text(seg_by->'columns') t(x))::text[], "
+					 " seg_by->>'message', "
+					 " (seg_by->>'confidence')::int "
+					 "FROM %s(%d) seg_by",
+					 ts_guc_default_segmentby_fn,
+					 ht->main_table_relid);
```

Comment:
i think we need to lock down this query either with explicit search_path or fully schema qualifiying all the things.
Injecting the fn name directly from the GUC is SQL injection and we should quote it as identifier before using in a query

Reply:
fixed